modules/core/src/main/scala/gem/Location.scala

```diff
@@ -0,0 +1,73 @@
+package gem
+
+import scalaz._, Scalaz._
+import scalaz.Ordering.{EQ, GT, LT}
+
+/** A sortable value used to indicate relative positions of a set of associated
+  * elements.  `Location`s may be thought of as lists of arbitrary integers
+  * which are be compared element by element such that the first pair of values
+  * that differ determine the ordering of the `Location`s as a whole.  If one
+  * `Location` is a proper prefex of another, then it sorts ahead of the other
+  * `Location`.
+  */
+sealed abstract case class Location(toNel: NonEmptyList[Int]) {
+
+  // Use the sketchy sealed abstract case class technique to control the
+  // NonEmptyList values that are passed to the constructor.  We want to
+  // ensure that trailing Int.MinValue is always trimmed so that EQ and ==
+  // agree.
+
+  def toIList: IList[Int] =
+    toNel.list
+
+  def toList: List[Int] =
+    toIList.toList
+
+  override def toString: String =
+    toList.mkString("{", ",", "}")
+}
+
+object Location {
+  private def trim(is: Seq[Int]): IList[Int] =
+    trim(IList(is: _*))
+
+  // Remove trailing Int.MinValue so that == agrees with EQ.
+  private def trim(is: IList[Int]): IList[Int] =
+    is.dropRightWhile(_ == Int.MinValue)
+
+  def apply(h: Int, t: Int*): Location =
+    new Location(NonEmptyList.nel(h, trim(t))) {}
+
+  def fromList(is: List[Int]): Option[Location] =
```

Comment:
You can define this as `def from[F[_]: Foldable](fi: F[Int])` so you don't need the `List` and `IList` variants, or just hang onto the passed foldable if you're generic in `F` as suggested above.